### ISSUE TITLE
FileSystemEvent.isDirectory property returns wrong value

### ISSUE THREAD

**Issue by sgrekhov** [dart-lang/sdk#30359]
The code is the following:
```dart
import "dart:io";
import "dart:async";

main() {
  Directory dir = new Directory(Directory.systemTemp.path + Platform.pathSeparator + "test-dir");
  dir.createSync();
  StreamSubscription s = dir.watch().listen((FileSystemEvent event) {
    print(event);
    print(event.isDirectory);
  });

  Directory d = dir.createTempSync();
  d.delete().then((_) {}).timeout(new Duration(seconds: 1)).then((_) {
    s.cancel().then((_) {
      dir.delete(recursive: true);
    });
  });
}
```
This prints for me:
```
FileSystemCreateEvent('C:\Users\sgrekhov\AppData\Local\Temp\test-dir\e77feb7c-7c24-11e7-b777-bcaec534d2ad')
false
FileSystemDeleteEvent('C:\Users\sgrekhov\AppData\Local\Temp\test-dir\e77feb7c-7c24-11e7-b777-bcaec534d2ad')
false
```
`event.isDirectory` here must be `true`

Another example:
```dart
import "dart:io";
import "dart:async";
main() {
  Directory dir = new Directory(Directory.systemTemp.path + Platform.pathSeparator + "test-dir");
  dir.createSync();
  StreamSubscription s = dir.watch().listen((FileSystemEvent event) {
    print(event);
    print(event.isDirectory);
  });
  dir.createTemp().then((Directory d) {
    d.delete().timeout(new Duration(seconds: 1)).then((_) {
      s.cancel().then((_) {
        dir.delete(recursive: true);
      });
    });
  });
}
```
This produces:
```
FileSystemCreateEvent('C:\Users\sgrekhov\AppData\Local\Temp\test-dir\5e409e53-7c25-11e7-b777-bcaec534d2ad')
true
FileSystemDeleteEvent('C:\Users\sgrekhov\AppData\Local\Temp\test-dir\5e409e53-7c25-11e7-b777-bcaec534d2ad')
false
```
`event.isDirectory` here must be `true` in the second case.
Tested on Dart VM version: 1.25.0-dev.9.0 (Tue Aug 01 01:52:00 2017) on "windows_x64"

**Comment by zanderso** [comment]
On Windows, the file system watcher is implemented with [`ReadDirectoryChangesW`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365465(v=vs.85).aspx). From `ReadDirectoryChangesW` we get a stream of [`FILE_NOTIFY_INFORMATION`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364391(v=vs.85).aspx) structures. The `FILE_NOTIFY_INFORMATION` structures do not indicate the type of file that the event is for. Therefore, for delete events, we no longer have access to the type of the file.

Given this limitation, I don't think there's a nice way to address this issue. Instead, I'm going to a documentation issue. In particular, [`FileSystemEvent.isDirectory`](https://api.dartlang.org/stable/1.24.2/dart-io/FileSystemEvent/isDirectory.html) should be updated to indicate that it will always return false for delete events on Windows even if the event was for a directory.

**Comment by zanderso** [comment]
https://codereview.chromium.org/2997573002/

**Comment by sgrekhov** [comment]
@zanderso please look at the following
```dart
import "dart:io";
import "dart:async";

main() {
  Directory dir = new Directory(Directory.systemTemp.path + Platform.pathSeparator + "test-dir");
  dir.createSync();
  StreamSubscription s = dir.watch().listen((FileSystemEvent event) {
    print(event);
    print(event.isDirectory);
  });

  Directory d = dir.createTempSync();
  d.delete().then((_) {}).timeout(new Duration(seconds: 1)).then((_) {
    s.cancel().then((_) {
      dir.delete(recursive: true).then((_) {
        print("Deleted");
      });
    });
  });
}
```
This prints 
```
FileSystemCreateEvent('C:\Users\sgrekhov\AppData\Local\Temp\test-dir\d6d89476-7cb3-11e7-b777-bcaec534d2ad')
false
FileSystemDeleteEvent('C:\Users\sgrekhov\AppData\Local\Temp\test-dir\d6d89476-7cb3-11e7-b777-bcaec534d2ad')
false
Deleted
```
Note, that by the time the both (create and delete) events have arrived the entity is not deleted yet but `isDirectory` is `false` for the both events anyway. So, I think, suggested explanation doesn't cover this case. In particular (look to the code examples from original issue description), seems, that if temp directory is created by `Directory.createTemp()` then `isDirectory` is true, but when it crreated by `Directory.createTempSync()` then `isDirectory` for `CREATE` event is false

**Comment by zanderso** [comment]
Note that Dart Isolates are single threaded, and that there are no ordering guarantees for asynchronous callbacks. Thus, the listen callback for the create event in all of these examples is not guaranteed to execute before the file is deleted. If it does, then it is only by luck. You'd need some additional synchronization (like a Completer completing in the listen callback) to make it happen every time.

tl;dr: In all of these examples, the listen callback is racing with the delete operation.
